src/dmd/dsymbol.d

```diff
@@ -663,8 +663,25 @@ extern (C++) class Dsymbol : ASTNode
         // Fill in comp[] and compute length of final result
         size_t length = 0;
         int i;
-        for (Dsymbol p = this; p; p = p.parent)
+        for (Dsymbol p = this, pparent; p; p = pparent)
         {
+            Dsymbol sym;
+            pparent = p.parent;
+            if (pparent)
+            {
+                if (auto ti = pparent.isTemplateInstance())
+                    if (auto ident = p.getIdent())
+                        if (ident is ti.name)
+                            if (Dsymbol.oneMembers(ti.members, &sym, ident) && sym is p)
+                                continue;
+
+                if (auto td = pparent.isTemplateDeclaration())
```

Comment:
Since the logic is duplicated, I suggest moving the check in a function that receives a ScopeDsymbol (or a template function that receives a template instance or a template declaration) so that we save some lines of code.